### PR TITLE
Url arbitration

### PR DESCRIPTION
**Pull request by steventux**
This is the initial step to internalizing url arbitration in the publishing API. 
The aim of this work is to continue to write downstream url arbiter records whilst populating internal publishing API path reservations.
There are changes to gds-api-adapters which relate to this pull request.
Pact tests are expected to break with these changes because of unexpected status codes. The update to the pact is here https://github.com/alphagov/gds-api-adapters/pull/374 The failing pact tests are an expected side effect because of changing error codes returned by the API.

### CODE REVIEW

File: app/commands/reserve_url.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module Commands
+  class ReserveUrl < BaseCommand
+    def call
+      UrlReservation.reserve_base_path!(base_path, payload[:publishing_app])
```
Comment (danielroseman) [inline app/commands/reserve_url.rb]: This needs to call `Adapters::UrlArbiter` as well.

File: app/controllers/v2/url_reservations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class UrlReservationsController < ApplicationController
```
Comment (danielroseman) [inline app/controllers/v2/url_reservations_controller.rb]: Don't think we need this controller. It's not linked from the routes, in any case.

File: spec/commands/reserve_url_spec.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+RSpec.describe Commands::ReserveUrl do
+  describe "call" do
+    context "with a new base_path" do
+      let(:payload) {
+        { base_path: "/foo", publishing_app: "Foo" }
+      }
+
+      it "successfully reserves the path" do
+        expect(described_class.call(payload)).to be_a Commands::Success
```
Comment (danielroseman) [inline spec/commands/reserve_url_spec.rb]: needs expectation on calling url-arbiter.

**Comment by steventux** [comment]
@danielroseman  thanks for reviewing, updated accordingly.

File: spec/support/request_helpers/endpoint_behaviour.rb [hunk 4]
```diff
@@ -87,6 +87,38 @@ def accepts_root_path
...
+
+          expect(PathReservation.count).to eq(1)
+          expect(PathReservation.first.base_path).to eq(base_path)
+          expect(PathReservation.first.publishing_app).to eq(content_item[:publishing_app])
+        end
+      end
+
+      context "base_path has already been registered" do
+        it "should be successful if the publishing app matches" do
+          expect{ do_request }.to change(PathReservation, :count).by(1)
```
Comment (danielroseman) [inline spec/support/request_helpers/endpoint_behaviour.rb]: This should be testing against an existing PathReservation, and the count should not change.

**Pull request merged**